File: leet-code/interviews.school/04-sorting/merge-intervals.cpp
// https://leetcode.com/problems/merge-intervals/

#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> merge(std::vector<std::vector<int>>& intervals) {
        std::vector<std::vector<int>> result;
        
        // sort the numbers
        sort(intervals.begin(), intervals.end());
        std::vector<std::vector<int>>::const_iterator it = intervals.cbegin();

        while (it != intervals.cend()) {
            std::vector<int> merged = *it++;

            while (it != intervals.cend() && merged[1] >= (*it)[0]) {
                merged[1] = std::max(merged[1], (*it)[1]);
                it++;
            }

            result.push_back(merged);
        }

        return result;
    }
};

File: leetcode/educative.io/04-merge-intervals/merge-intervals.go
// https://leetcode.com/problems/merge-intervals/

package main

import "sort"

func merge(intervals [][]int) [][]int {
	var result [][]int

	// sort array
	sort.Slice(intervals[:], func(i, j int) bool {
		for a := range intervals[i] {
			if intervals[i][a] == intervals[j][a] {
				continue
			}
			return intervals[i][a] < intervals[j][a]
		}
		return false
	})

	for i := 0; i < len(intervals); {
		merged := intervals[i]
		i++

		for i < len(intervals) && merged[1] >= intervals[i][0] {
			if intervals[i][1] > merged[1] {
				merged[1] = intervals[i][1]
			}
			i++
		}

		result = append(result, merged)
	}

	return result
}
